### DISCUSSION TITLE
cycle-free graph violated: partition depends on itself

### DISCUSSION THREAD

**Discussion by ghost** [Xilinx/finn#381] [Q&A]
Hello, I am trying to implement a Lenet-5 quantized network by using the dev branch of FINN based on the cnv_end2end_example.ipynb tutorial. As soon as I try to do the partitioning by CreateDataflowPartition transformation to create a "dataflow partition" in this graph, I got the following  AssertionError: 
![image](https://user-images.githubusercontent.com/45943762/136215649-d6b6ac1b-71b4-4769-90ac-9d73bbf10971.png)
Could you please help me or introduce any hands-on tutorial that gives me better insight to fix the problem?

Bests,

**Comment by maltanar** [comment]
Hi Mahdi, when you see this error during the dataflow partitioning stage, it's usually because there are some "leftover layers" that the FINN compiler wasn't able to convert to HLS, interspersed with the HLS layers. The `CreateDataflowPartition` expects there to be a single contiguous partition of HLS layers, e.g.:

* Good: `(non-HLS layer) -> (non-HLS layer) -> (HLS layer) -> (HLS layer) -> (non-HLS layer)`
* Bad:  `(non-HLS layer) -> (HLS layer) -> (non-HLS layer) -> (HLS layer) -> (non-HLS layer)` (needs several dataflow partitions)

You can take a look at the model prior to dataflow partitioning using Netron and see if you can spot the problem. If you find out which layers haven't been converted to HLS, we can have a look and see if you can convert those layers as well by calling more graph transformations.

**Reply by ghost** [reply]
Dear Yaman,
Thanks for your reply. Here is the list of my nodes in my .onnx file after Streamlining transform before doing data partitioning. 
{'MultiThreshold': 5, 'Transpose': 4, 'Im2Col': 2, 'MatMul': 5, 'MaxPoolNHWC': 1, 'MaxPool': 1, 'Reshape': 1, 'TopK': 1}
I have also attached the visualised graph below. I couldn't see any problem in my transferred model. Am I missing any points?
![finn_lenet_2_streamlined onnx](https://user-images.githubusercontent.com/45943762/136347071-d74636c6-edda-446f-b1e9-4b61a9835577.png)

**Reply by heborras** [reply]
Hi @mahditaheri,
thank you for sharing the graph.
There are indeed a few nodes, which did not get converted to their FINN counterparts or shouldn't be there. These are the second and third Transpose nodes, which should not exist at this point in the graph. And the MaxPool node towards the end of the graph might be an issue, but it's hard to tell from the streamlined model alone. Please take a look at the streamlined network (`2_step_streamline.onnx`) in the attached zip file to get an impression for how the network should look after streamlining: [CNV_W1A1_streamlined_and_hls.zip](https://github.com/Xilinx/finn/files/7312052/CNV_W1A1_streamlined_and_hls.zip)
 Note that this network uses `XnorPopcountMatMul` nodes instead of `MatMul` nodes, because it is a binary network.

As to what to do about these nodes: 

- The second and third transpose node you could probably manually remove, as these should have no impact on the network. 
- As for the MaxPool node, here one would actually need to look at what happens after the HLS conversion (still before the `CreateDataflowPartition` transformation). This conversion happens in the next step, where FINN nodes get converted to HLS nodes, these should then have the domain/module called `finn.custom_op.fpgadataflow` in the onnx graph. The "3_step_convert_to_hls.onnx" file in the .zip file shows a corresponding example. This would also be the step in the FINN toolchain, which @maltanar was talking about before. 
Could you maybe share what that model looks like after the HLS conversion or even better the specific onnx file?

**Reply by julienmas** [reply]
Hi @HenniOVP , how do you manually remove a node?
